src/Avalonia.Controls/DateTimePickers/DateTimeFormatter.cs

```diff
@@ -0,0 +1,335 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Avalonia.Controls
+{
+    /// <summary>
+    /// Formats a DateTimeOffset by the specified pattern
+    /// Based on, but not replica, the Windows.Globalization.DateTimeFormatting.DateTimeFormatter
+    /// https://docs.microsoft.com/en-us/uwp/api/windows.globalization.datetimeformatting.datetimeformatter?view=winrt-19041
+    /// <para>
+    /// Currently only the Gregorian Calendar is supported, and regions and languages are not supported. 
+    /// Formatting timezones is also not supported, that's a large task
+    /// </para>
+    /// <para>
+    /// /// Most formats used in UWP/WinUI are compatible here, but not all
+    /// This DateTimeFormatter will also work with TimeSpans (only patterns though),
+    /// in addition to the default DateTime/DateTimeOffset
+    /// </para>
+    /// <para>
+    /// Formats are broken down into Patterns and Templates, which are "complete" patterns
+    /// If a Template is used, only 1 may be provided and cannot be mixed with anything else
+    /// Multiple patterns can be specified, and can be mixed with other text. All patterns
+    /// must be enclosed in curly braces {}, e.g. {dayofweek}, or in xaml "{}{dayofweek}"
+    /// NOTE: Formats and Templates are case sensitive
+    /// </para>
+    /// </summary>
+    public sealed class DateTimeFormatter
```

Comment:
I have huge reservations regarding this class... If we can help it, we should just use the built in DateTime formatting tools in the standard library...

Reply:
Understandable. I took a stab at porting what I could of this because it provided a little better formatting control over regular date/time formatting and it brings the usage of the Date/Time pickers more inline with WinUI. This could be reverted, though I'm not sure how hard that would be

Reply:
Keep in mind WinRT and therefore UWP/WinUI does NOT have access to the .net framework world. This is so it can support C++ and other frameworks. A side effect of this I think is visible here and should be corrected for Avalonia. You shouldn't re-invent the wheel as @jmacato stated. Instead, the .net libraries should be used for this which is possible in Avalonia but not in WinUI.